### PR TITLE
Add TELEMETRY_WRITABLE build option to enable telemetry write mode (default read-only)

### PR DESCRIPTION
**Pull request by seiferteric**
Add build flag TELEMETRY_WRITABLE. When set to "y" it will add a go build flag in the telemetry build that will enable telemetry write mode to allow configuration via gNMI Set RPC as well as operations via the gNOI RPC's. The default for TELEMETRY_WRITABLE is unset in which case telemetry is read-only. In read-only mode the Set RPC and all gNOI RPC's are disabled and will return an "Unsupported" error when called.

### CODE REVIEW

File: Makefile.work [hunk 1]
```diff
@@ -183,6 +183,7 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            SONIC_ENABLE_SYSTEM_TELEMETRY=$(ENABLE_SYSTEM_TELEMETRY) \
                            SONIC_ENABLE_RESTAPI=$(ENABLE_RESTAPI) \
                            EXTRA_JESSIE_TARGETS=$(EXTRA_JESSIE_TARGETS) \
+			   TELEMETRY_READONLY=$(TELEMETRY_READONLY) \
```
Comment (jeff-yin) [inline Makefile.work]: fix indent

**Comment by renukamanavalan** [comment]
As per comment in [PR #27](https://github.com/Azure/sonic-telemetry/pull/27), please plan to configure rw mode of a path mounted into this docker, instead of a separate file.

Or you might play with some config file entry that is local to telemetry.

**Comment by seiferteric** [comment]
The method has been changed to using a go build flag /constraint to set a constant in the telemetry binary.

File: Makefile.work [hunk 2]
```diff
@@ -183,6 +183,7 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            SONIC_ENABLE_SYSTEM_TELEMETRY=$(ENABLE_SYSTEM_TELEMETRY) \
                            SONIC_ENABLE_RESTAPI=$(ENABLE_RESTAPI) \
                            EXTRA_JESSIE_TARGETS=$(EXTRA_JESSIE_TARGETS) \
+                           TELEMETRY_READWRITE=$(TELEMETRY_READWRITE) \
```
Comment (renukamanavalan) [inline Makefile.work]: Please update the 'Supported parameters:' list at the top.
Also update rules/config, with the added lines under comments, as it is not used by default.

File: slave.mk [hunk 3]
```diff
@@ -209,6 +209,7 @@ $(info "BLDENV"                          : "$(BLDENV)")
 $(info "VS_PREPARE_MEM"                  : "$(VS_PREPARE_MEM)")
 $(info "ENABLE_SFLOW"                    : "$(ENABLE_SFLOW)")
 $(info "ENABLE_NAT"                      : "$(ENABLE_NAT)")
+$(info "TELEMETRY_READWRITE"             : "$(TELEMETRY_READWRITE)")
```
Comment (renukamanavalan) [inline slave.mk]: Have you tested this change? 
Setting this flag to Yes, while calling Make, did it propagate to Make of Telemetry?

File: Makefile.work (hunk 2 above)
Comment (seiferteric) [reply Makefile.work]: Okay, dones.

File: slave.mk (hunk 3 above)
Comment (seiferteric) [reply slave.mk]: I have tested like this:
`TELEMETRY_READWRITE=y make target/debs/stretch/sonic-telemetry_0.1_amd64.deb` and loaded to a VM switch and confirmed that gnmi_set is allowed.

Then:
`make target/debs/stretch/sonic-telemetry_0.1_amd64.deb` and loaded to VM switch and confirmed that gnmi_set is not allowed with "Unimplemented" and error message: "Telemetry is in read-only mode"

**Comment by lguohan** [comment]
can you resolve the conflict?

**Comment by seiferteric** [comment]
> can you resolve the conflict?

Okay, done.

File: Makefile.work [hunk 4]
```diff
@@ -192,6 +195,7 @@ SONIC_BUILD_INSTRUCTION :=  make \
...
```
Comment (lguohan) [inline Makefile.work]: this does not work. SONIC_TELEMETRY_READWRITE=$(TELEMETRY_READWRITE)

File: Makefile.work [hunk 5]
```diff
@@ -31,6 +31,9 @@
 #  *                 Default: yes
 #  *                 Values:  yes, no
 #  * KERNEL_PROCURE_METHOD: Specifying method of obtaining kernel Debian package: download or build
+#  * TELEMETRY_READWRITE: Enable write/config operations via the gNMI interface.
```
Comment (lguohan) [inline Makefile.work]: change to SONIC_TELEMETRY_READWRITE

Comment (seiferteric) [reply Makefile.work]: Okay, I have added the SONIC_ prefix to the telemetry readonly flag.

File: Makefile.work [hunk 6]
```diff
@@ -31,6 +31,9 @@
 #  *                 Default: yes
 #  *                 Values:  yes, no
 #  * KERNEL_PROCURE_METHOD: Specifying method of obtaining kernel Debian package: download or build
+#  * SONIC_TELEMETRY_READWRITE: Enable write/config operations via the gNMI interface.
```
Comment (jleveque) [inline Makefile.work:34]: Suggest rename to `SONIC_TELEMETRY_WRITABLE`, as enabling/disabling only modifies the ability to write, it doesn't change the ability to read.

Comment (seiferteric) [reply Makefile.work:34]: Sorry I missed this comment, @renukamanavalan @lguohan what do you think, is the name okay? or should it be changed to SONIC_TELEMETRY_WRITABLE? Also, is there anything else needed to get this merged?

Comment (renukamanavalan) [reply Makefile.work:35]: Let us take Joe's suggestion.

Comment (renukamanavalan) [reply Makefile.work:35]: The rest looks good.

Comment (seiferteric) [reply Makefile.work]: Okay I have changed it. I had to make the same change in telemetry as well, see: https://github.com/Azure/sonic-telemetry/pull/34

**Comment by renukamanavalan** [comment]
retest vs please.

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by renukamanavalan** [comment]
retest vs please

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by renukamanavalan** [comment]
retest mellanox please

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by renukamanavalan** [comment]
retest mellanox please

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by renukamanavalan** [comment]
retest mellanox please

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by ben-gale** [comment]
Can this one be merged now?

File: Makefile.work [hunk 7]
```diff
@@ -200,6 +203,8 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            HTTPS_PROXY=$(https_proxy) \
                            SONIC_ENABLE_SYSTEM_TELEMETRY=$(ENABLE_SYSTEM_TELEMETRY) \
                            SONIC_ENABLE_RESTAPI=$(ENABLE_RESTAPI) \
+                           EXTRA_JESSIE_TARGETS=$(EXTRA_JESSIE_TARGETS) \
```
Comment (renukamanavalan) [inline Makefile.work:206]: Why do we have this EXTRA_JESSIE_TARGETS ?
Is this a side effect of resolving conflict?

Comment (seiferteric) [reply Makefile.work]: Hmm, it seems so. I have removed it now. Thanks.

**Comment by renukamanavalan** [comment]
The PR has changed a lot, that the original message in the PR "_Add build flag TELEMETRY_READONLY ..._" is no longer relevant. Can you please update ?

As part of that please add, who/how this flag is consumed, to toggle gNMI i/f between RO/RW during build time.

**Comment by seiferteric** [comment]
@renukamanavalan Okay I have updated the comment and title.

File: slave.mk [hunk 8]
```diff
@@ -230,6 +230,8 @@ $(info "BLDENV"                          : "$(BLDENV)")
 $(info "VS_PREPARE_MEM"                  : "$(VS_PREPARE_MEM)")
 $(info "ENABLE_SFLOW"                    : "$(ENABLE_SFLOW)")
 $(info "ENABLE_NAT"                      : "$(ENABLE_NAT)")
+$(info "SONIC_TELEMETRY_WRITABLE"       : "$(SONIC_TELEMETRY_WRITABLE)")
```
Comment (jleveque) [inline slave.mk:233]: Keeping in line with the other options, I suggest removing the `SONIC_` prefix from the user-facing option name.

File: rules/config [hunk 9]
```diff
@@ -131,6 +131,9 @@ ENABLE_RESTAPI = n
 # ENABLE_NAT - build docker-sonic-nat for nat support
 ENABLE_NAT = y
 
+# SONIC_TELEMETRY_WRITABLE - Enable write/config operations via the gNMI interface.
+# Uncomment to enable:
+# SONIC_TELEMETRY_WRITABLE = y
```
Comment (jleveque) [inline rules/config:136]: Keeping in line with the other options, I suggest removing the `SONIC_` prefix from the user-facing option name.

Comment (seiferteric) [reply rules/config:136]: I previously had it without SONIC_ prefix and was told to add the SONIC_ prefix

Comment (jleveque) [reply rules/config:136]: Sorry. It seems like there must have been some confusion. For the user-facing option name, it seems redundant and a waste of screen space to prefix them all with "SONIC_". It appears as though some people haven't been consistent with this, however. We'll need to go through and clean some up in the future.

Comment (renukamanavalan) [reply rules/config:136]: Joe would it be OK to drop SONIC_ prefix from both internal & user-facing  ? 
This would make it cleaner and less error-prone.

Comment (jleveque) [reply rules/config:136]: Yes. I don't see a problem with removing it from both, unless other internal processing would need to be done before setting the internal variable, in which case, different names would make sense.

Comment (renukamanavalan) [reply rules/config:136]: @seiferteric , can we get this resolved?

Comment (seiferteric) [reply rules/config]: Okay done, also updated the corresponding telemetry PR https://github.com/Azure/sonic-telemetry/pull/34

**Comment by renukamanavalan** [comment]
retest mellanox please

**Comment by seiferteric** [comment]
> retest mellanox please

@renukamanavalan What do you mean by retest mellanox?

**Comment by renukamanavalan** [comment]
retest xyz please -- will restart the xyz test.

mellanox failed previously, due to some unreachability issue. So I restarted it.

**Comment by renukamanavalan** [comment]
Aborted by Danny Allen; restarting vs test.

**Comment by renukamanavalan** [comment]
retest vsimage please.

**Comment by renukamanavalan** [comment]
retest vsimage please

**Pull request merged**